### ISSUE TITLE
"Save failed" when removing description

### ISSUE THREAD

**Issue by adrian17** [metabase/metabase#4738]
Version: 0.23.1 running on Ubuntu 16
Internal DB: reproduced on H2 and MySQL.

Steps:

- add a description to question/dashboard
- save
- clear the description by clearing the text field
- save
- "save failed" appears.

Server-side error message:

    {:errors {:description "value may be nil, or if non-nil, value must be a non-blank string."}}

**Comment by mazameli** [comment]
Yeah, was able to reproduce this on master.

**Comment by adrian17** [comment]
I just encountered a similar issue:

- made the first dashboard
- went to "Data reference"
- clicked "get started"
- chose the dashboard, filled the first two text fields, clicked "save"

error:

    {:errors {:show_in_getting_started "value may be nil, or if non-nil, value must be a non-blank string."}}

Looks like both are blocked by API definition: https://github.com/metabase/metabase/blob/master/src/metabase/api/dashboard.clj#L76

Another one is in `reference/databases/`: In database description, all four fields are mandatory, and there are no user-visible errors when you try saving without any of them. This is especially confusing, since the first field in database/table editor has a default value (name of DB/table), which implies you don't need to insert it manually.

**Comment by tlrobinson** [comment]
`su/NonBlankString` strikes again.

**Comment by camsaul** [comment]
@tlrobinson it's `(s/maybe su/NonBlankString)` (hence the `value may be nil` in the error message)

to clear out the text field you have to pass `nil`, because you can't have an empty string value for description

**Comment by salsakran** [comment]
@tlrobinson + @camsaul is it useful to distinguish between "" and nil in this situation? 

Do we have a general principled stance on this codebase wide? This seems like the kind of question we should have a concrete position on rather than argue out API response endpoints one at a time.

**Comment by camsaul** [comment]
Yeah that's the pattern we've had all along across the whole API. Unsetting something in the API works by setting it to `nil`.

Otherwise it's subject to validation to make sure it's a valid value for that key. For example an email or URL gets validated against a regex, and a blank string would fail.

**Comment by tlrobinson** [comment]
The validation was added after the fact, which is why stuff keeps breaking.

**Comment by camsaul** [comment]
yeah we're a little better about validating stuff after I made the tweaks so the `defendpoint` macro could validate params with `schema` last year but the general underlying way things worked didn't change

**Comment by camsaul** [comment]
i think we just need to make the text widget component return `null` if it's empty and we'll be g2g

**Comment by camsaul** [comment]
Fixed by #4757

**Comment by vvaezian** [comment]
I'm getting this error in 0.33.3.

**Comment by flamber** [comment]
@vvaezian But how/when? Clearing question/dashboard description or reference description? Since it's a couple of years old, then it might be better with a new issue.

**Comment by vvaezian** [comment]
It was a table description. I tried both in data model and the reference section. 
Sure. I'll create a new issue.
